modules/python/python-http-header.c

```diff
@@ -0,0 +1,383 @@
+                evt_tag_str("exception", _py_format_exception_text(buf, sizeof(buf))));
+      _py_finish_exception_handling();
+
+      goto exit;
+    }
+
+  py_args = _py_create_arg_dict(self->options);
+  if (!py_args)
+    {
+      gchar buf[256];
+
+      msg_error("Error creating argument dictionary",
+                evt_tag_str("class", self->class),
+                evt_tag_str("exception", _py_format_exception_text(buf, sizeof(buf))));
+      _py_finish_exception_handling();
+
+      goto exit;
+    }
+
+  self->py.instance = _py_invoke_function(self->py.class, py_args, self->class, self->super.name);
+  if (!self->py.instance)
+    {
+      gchar buf[256];
+
+      msg_error("Error instantiating Python class",
+                evt_tag_str("class", self->class),
+                evt_tag_str("exception", _py_format_exception_text(buf, sizeof(buf))));
+      _py_finish_exception_handling();
+
+      goto exit;
+    }
+
+  self->py.get_headers = _py_get_attr_or_null(self->py.instance, "get_headers");
+  if (!self->py.get_headers)
+    {
+      msg_error("Error initializing plugin, required method not found",
+                evt_tag_str("class", self->class),
+                evt_tag_str("method", "get_headers"));
+    }
+
+exit:
+  Py_XDECREF(py_args);
+  return self->py.get_headers != NULL;
+}
+
+static void
+_py_detach_bindings(PythonHttpHeaderPlugin *self)
+{
+  Py_CLEAR(self->py.class);
+  Py_CLEAR(self->py.instance);
+  Py_CLEAR(self->py.get_headers);
+}
+
+static void
+_append_str_to_list(gpointer data, gpointer user_data)
+{
+  List *list = (List *) user_data;
+  list_append(list, data);
+}
+
+static void
+_append_headers(PythonHttpHeaderPlugin *self, HttpHeaderRequestSignalData *data)
+{
+  PyObject *py_ret = NULL,
+            *py_list = NULL,
+             *py_args = NULL,
+              *py_ret_list = NULL;
+
+  time_t now = time(NULL);
+  if ((now - self->last_run) < self->timeout)
+    goto exit;
+
+  // The GIL also guards non-Python plugin struct members from concurrent changes below.
+  PyGILState_STATE gstate = PyGILState_Ensure();
+
+  self->timeout = 0;
+  g_list_free_full(self->last_headers, g_free);
+  self->last_headers = NULL;
+
+  py_list = _py_convert_list_to_pylist(data->request_headers);
+  if (!py_list)
+    goto cleanup;
+
+  py_args = Py_BuildValue("(sO)", data->request_body->str, py_list);
+  if (!py_args)
+    {
+      gchar buf[256];
+
+      msg_error("Error creating Python arguments",
+                evt_tag_str("class", self->class),
+                evt_tag_str("exception", _py_format_exception_text(buf, sizeof(buf))));
+      _py_finish_exception_handling();
+
+      goto cleanup;
+    }
+
+  py_ret = _py_invoke_function_with_args(self->py.get_headers, py_args, self->class, "_append_headers");
+  if (!py_ret || !PyArg_ParseTuple(py_ret, "iO", &self->timeout, &py_ret_list))
+    {
+      gchar buf[256];
+
+      msg_error("Invalid response returned by Python call",
```

Comment:
If there is a bug in the user code (let's say uncatched exception), then this error message is emitted, and the http destination emits the response as if no problem has happened.

I think the correct behavior would be that an error in the python http header plugin would cause error in the http destination, and nothing must be emitted: do a disconnect-reconnect cycle.

This might be challenging to implement, because for now we do not have the protocol to send success/failure through the signal-slot mechanism.

Reply:
It is intentional to not add a generic error handling mechanism to signal-slot model.

When a signal owner needs error handling, then it should be part of the interface and each signal-slot connections should implement its own error handling strategy (so in this specific case, we can extend `HttpHeaderRequestSignalData` with some error state).